docs/concurrency-adoption-guidelines.md

```diff
@@ -0,0 +1,201 @@
+
+These relaxed semantics for imported code are not implemented yet, but keep it in mind when working with C APIs from Swift and trying to adopt the `-warn-concurrency` mode today. Please file any issues you hit on [bugs.swift.org](https://bugs.swift.org/secure/Dashboard.jspa) so we can inform the development of these checking heuristics based on real issues you hit.
+
+### Custom Executors
+
+We expect that Swift Concurrency will allow custom executors in the future. A custom executor would allow the ability to run actors / tasks “on” such executor. It is possible that `EventLoop`s could become such executors, however the custom executor proposal has not been pitched yet.
+
+While we expect potential performance gains from using custom executors “on the same event loop” by avoiding asynchronous hops between calls to different actors, their introduction will not fundamentally change how NIO libraries are structured.
+
+The guidance here will evolve as Swift Evolution proposals for Custom Executors are proposed, but don’t hold off adopting Swift Concurrency until custom executors “land” - it is important to start adoption early. For most code we believe that the gains from adopting Swift Concurrency vastly outweigh the slight performance cost actor-hops might induce.
+
+
+### Reduce use of Swift NIO Futures as “Concurrency Library“
+
+Swift NIO currently provides a number of currency types for the Swift on Server ecosystem. Most notably `EventLoopFuture`s and `EventLoopPromise`s, that are used widely for asynchronous results. While the SSWG recommended using those at the API level in the past for easier interplay of server libraries, we advise to deprecate or remove such APIs once Swift 6 lands. The swift-server ecosystem should go all in on the structured concurrency features the languages provides. For this reason, it is crucial to provide async/await APIs today, to give your library users time to adopt the new APIs.
+
+Some NIO types will remain however in the public interfaces of Swift on server libraries. We expect that networking clients and servers continue to be initialized with `EventLoopGroup`s. The underlying transport mechanism (`NIOPosix` and `NIOTransportServices`) should become implementation details however and should not be exposed to library adopters.
+
+### NIO 3
+
+While subject to change, it is likely that Swift NIO will cut a 3.0 release in the months after Swift 6.0, at which point in time Swift will have enabled “full” `Sendable` checking.
+
+Do not expect NIO to suddenly become “more async”, NIO’s inherent design principles are about performing small tasks on the event loop and using Futures for any async operations. The design of NIO is not expected to change. It is crucial to its high performance networking design. Channel pipelines are not expected to become “async”.
```

Comment:
>The design of NIO is not expected to change. It is crucial to its high performance networking design.

This line seems to imply that `async` cannot be used without significant overhead, is this the case? If so, should we include guidance in this document about when _not_ to completely adopt `async`, as NIO has opted not to? Or maybe a note saying NIO is a special case due to its low level nature, and that most/all other libraries should try to become more `async`?

The reason I mention this is that the majority of the document focuses on encouraging adoption of `async`, but at the very end here we have a note about how NIO will largely not be doing so, which may be a bit contradictory. A little bit further clarification could help with that though.

Reply:
NIO is not doing so because of the concerns around the “bottom” of the `async` pyramid. `async`/`await` is built on top of a co-operative thread pool within the Swift runtime. This thread pool must not be blocked by any operation, because doing so will starve the pool and prevent further progress of other async tasks.

This poses a problem for I/O. At some point, someone somewhere has to block a thread waiting for more I/O events, either in an I/O syscall or in something like `epoll_wait`. This is how NIO works: each of the event loop threads ultimately blocks on `epoll_wait`. We can’t do that inside the cooperative thread pool, as to do so would starve it for other async tasks, so we’d have to do so on a different thread.

This would mean we’d have a thread-hop between each I/O operation and dispatching it onto the async/await pool. This is not acceptable for high performance I/O: the context switch for each I/O op is too expensive. As a result, until we can safely block a thread for ourselves (with concurrent executors) we simply cannot use the async/await pattern in NIO code.

The rule of thumb is: if you need the absolute fastest computation on data from the network, you write `ChannelHandler`s and do your I/O on event loops. If you can tolerate slightly less performance there, likely because your business logic is more substantial, then async/await will repay you in ease of use much more than the (small) amount it’ll cost you in perf.

Reply:
That all makes sense, thanks for the detailed answer! I think it may be beneficial then to reword the existing paragraph to include some of this context and to be bit more explanatory rather than declarative (i.e. "NIO's API will largely remain the same because..." rather than "don't expect NIO's API to change..."). This will also help library authors decide when they might need to use NIO's facilities more directly rather than higher level `async`/`await` ones.

Reply:
Yeah, happy to take some of the wording into the doc 👍